File: app/src/main/java/com/example/mz/udacitypopularmovies/ReviewAdapter.java
package com.example.mz.udacitypopularmovies;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.mz.udacitypopularmovies.data.ReviewEntry;

import java.util.ArrayList;
import java.util.Arrays;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by mz on 2017-02-04.
 */
public class ReviewAdapter extends RecyclerView.Adapter<ReviewAdapter.ReviewViewHolder> {

    private static final String LOG_TAG = ReviewAdapter.class.getSimpleName();

    private ArrayList<ReviewEntry> mReviews;

    public ReviewAdapter() {
    }

    public void setReviewData(ReviewEntry[] reviewData) {
        Log.i(LOG_TAG, "before setMovieData there is " + getItemCount() + " elements in a view");
        if (reviewData == null) {
            mReviews = null;
        } else if (mReviews == null){
            mReviews = new ArrayList<>(Arrays.asList(reviewData));
        }
        else {
           mReviews.addAll(Arrays.asList(reviewData));
        }
        notifyDataSetChanged();
        Log.i(LOG_TAG, "after setMovieData there is " + getItemCount() + " elements in a view");
    }

    @Override
    public ReviewViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        int layoutIdForGridItem = R.layout.review_details;
        LayoutInflater inflater = LayoutInflater.from(context);
        final boolean shouldAttachToParentImmediately = false;
        View view = inflater.inflate(layoutIdForGridItem, parent, shouldAttachToParentImmediately);
        return new ReviewViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ReviewViewHolder holder, int position) {
        holder.bind(position);
    }

    @Override
    public int getItemCount() {
        if (mReviews == null) {
            return 0;
        } else {
            return mReviews.size();
        }
    }

    public class ReviewViewHolder extends RecyclerView.ViewHolder {
        private final String LOG_TAG = ReviewViewHolder.class.getSimpleName();
        @BindView(R.id.review_author)
        TextView tv_reviewAuthor;
        @BindView(R.id.review_content)
        TextView tv_reviewContent;

        public ReviewViewHolder(View rootView) {
            super(rootView);
            ButterKnife.bind(this, rootView);
        }


        public void bind(int position) {
            if (mReviews == null) {
                Log.d(LOG_TAG, "Called bind with null mReviews");
                return;
            }
            ReviewEntry entry = mReviews.get(position);
            tv_reviewAuthor.setText(entry.author);
            tv_reviewContent.setText(entry.content);
        }
    }
}

File: app/src/main/java/com/mz/popmovies/ReviewAdapter.kt
package com.mz.popmovies

import android.util.Log
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import com.mz.popmovies.data.ReviewEntry
import com.mz.popmovies.databinding.ReviewDetailsBinding
import java.util.*

/**
 * Created by mz on 2017-02-04.
 */
class ReviewAdapter : RecyclerView.Adapter<ReviewAdapter.ViewHolder>() {
    private var mReviews: ArrayList<ReviewEntry>? = null
    lateinit var binding: ReviewDetailsBinding
    fun setReviewData(reviewData: Array<ReviewEntry>?) {
        Log.i(LOG_TAG, "before setReviewData there is $itemCount elements in a view")
        if (reviewData == null) {
            mReviews = null
        } else if (mReviews == null) {
            mReviews = ArrayList(reviewData.asList())
        } else {
            mReviews!!.addAll(reviewData.asList())
        }
        notifyDataSetChanged()
        Log.i(LOG_TAG, "after setReviewData there is $itemCount elements in a view")
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        binding = ReviewDetailsBinding.inflate(inflater, parent, false)
        return ViewHolder(binding.root)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(position)
    }

    override fun getItemCount(): Int = mReviews?.size ?: 0

    inner class ViewHolder(rootView: View) : RecyclerView.ViewHolder(rootView) {
        private val LOG_TAG = ViewHolder::class.java.simpleName

        fun bind(position: Int) {
            if (mReviews == null) {
                Log.d(LOG_TAG, "Called bind with null mReviews")
                return
            }
            val entry = mReviews!![position]
            binding.reviewAuthor.text = entry.author
            binding.reviewContent.text = entry.content
        }
    }

    companion object {
        private val LOG_TAG = ReviewAdapter::class.java.simpleName
    }
}